pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+	}
+	defer func() {
+		_ = rdr.Close()
+	}()
+
+	meta, err := decodeMeta(rdr, tmpMetaFilename)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+
+	// Upload meta file so block is considered complete
+	if err := c.uploadMeta(ctx, logger, meta, blockID, block.MetaFilename, userBkt); err != nil {
+		return err
+	}
+
+	if err := userBkt.Delete(ctx, path.Join(blockID.String(), tmpMetaFilename)); err != nil {
+		level.Warn(logger).Log("msg", fmt.Sprintf(
+			"failed to delete %s from block in object storage", tmpMetaFilename), "err", err)
+		return nil
+	}
+
+	level.Debug(logger).Log("msg", "successfully completed block upload")
+	return nil
+}
+
+// sanitizeMeta sanitizes and validates a metadata.Meta object. If a validation error occurs, an error
+// message gets returned, otherwise an empty string.
+func (c *MultitenantCompactor) sanitizeMeta(logger log.Logger, tenantID string, blockID ulid.ULID,
+	meta *metadata.Meta) string {
+	meta.ULID = blockID
+
+	for l, v := range meta.Thanos.Labels {
+		switch l {
+		// Preserve these labels
+		case mimir_tsdb.CompactorShardIDExternalLabel:
+		// Remove unused labels
+		case mimir_tsdb.DeprecatedTenantIDExternalLabel, mimir_tsdb.DeprecatedIngesterIDExternalLabel, mimir_tsdb.DeprecatedShardIDExternalLabel:
+			level.Debug(logger).Log("msg", fmt.Sprintf("removing unused external label from %s", block.MetaFilename),
+				"label", l, "value", v)
+			delete(meta.Thanos.Labels, l)
+		default:
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting unsupported external label in %s", block.MetaFilename),
+				"label", l)
+			return fmt.Sprintf("unsupported external label in %s: %s", block.MetaFilename, l)
+		}
+	}
+
+	for _, f := range meta.Thanos.Files {
+		if f.RelPath == block.MetaFilename {
+			continue
+		}
+
+		if !rePath.MatchString(f.RelPath) {
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting file with invalid path in %s", block.MetaFilename),
+				"file", f.RelPath)
+			return fmt.Sprintf("file with invalid path in %s: %s", block.MetaFilename,
+				f.RelPath)
+		}
+
+		if f.SizeBytes <= 0 {
+			level.Warn(logger).Log("msg", fmt.Sprintf("rejecting file with invalid size in %s", block.MetaFilename),
+				"file", f.RelPath)
+			return fmt.Sprintf("file with invalid size in %s: %s", block.MetaFilename,
+				f.RelPath)
+		}
+	}
+
+	// validate minTime/maxTime
+	// basic sanity check
+	if meta.MinTime < 0 || meta.MaxTime < 0 || meta.MaxTime < meta.MinTime {
+		level.Warn(logger).Log("msg", fmt.Sprintf("invalid minTime/maxTime in %s", block.MetaFilename),
+			"minTime", meta.MinTime, "maxTime", meta.MaxTime)
+		return fmt.Sprintf("invalid minTime/maxTime in %s: minTime=%d, maxTime=%d",
+			block.MetaFilename, meta.MinTime, meta.MaxTime)
+	}
+	// validate that times are in the past
+	now := time.Now()
+	if time.UnixMilli(meta.MinTime).After(now) || time.UnixMilli(meta.MaxTime).After(now) {
```

Comment:
We can make this more readable by:

```suggestion
	now := time.Now().UnixMilli()
	if meta.MinTime > now || meta.MaxTime > now {
```